src/workload.rs

```diff
@@ -1248,42 +1251,72 @@ impl WorkloadStore {
                 debug!("failed to fetch workload for {}", ep.address);
                 return None
             };
+            let wl = wl.read().unwrap();
             // If endpoint overrides the target port, use that instead
             let target_port = ep.port.get(&addr.port()).unwrap_or(target_port);
             let mut us = Upstream {
                 workload: wl.to_owned(),
                 port: *target_port,
             };
-            Self::set_gateway_address(&mut us, hbone_port);
-            debug!("found upstream {} from VIP {}", us, addr.ip());
-            return Some(us);
+            match self.set_gateway_address(&mut us, hbone_port) {
+                Ok(_) => {
+                    debug!("found upstream {} from VIP {}", us, addr.ip());
+                    return Some(us);
+                }
+                Err(e) => {
+                    debug!("failed to set gateway address for upstream: {}", e);
+                    return None;
+                }
+            };
         }
         if let Some(wl) = self.workloads.get(&network_addr(network, addr.ip())) {
+            let wl = wl.read().unwrap();
             let mut us = Upstream {
                 workload: wl.to_owned(),
                 port: addr.port(),
             };
-            Self::set_gateway_address(&mut us, hbone_port);
-            debug!("found upstream: {}", us);
-            return Some(us);
+            match self.set_gateway_address(&mut us, hbone_port) {
+                Ok(_) => {
+                    debug!("found upstream {}", us);
+                    return Some(us);
+                }
+                Err(e) => {
+                    debug!("failed to set gateway address for upstream: {}", e);
+                    return None;
+                }
+            };
         }
         None
     }
 
-    fn set_gateway_address(us: &mut Upstream, hbone_port: u16) {
+    fn set_gateway_address(&self, us: &mut Upstream, hbone_port: u16) -> anyhow::Result<()> {
         if us.workload.gateway_address.is_none() {
             us.workload.gateway_address = Some(match us.workload.protocol {
                 Protocol::HBONE => {
                     let ip = us
                         .workload
-                        .waypoint_svc_ip_address()
-                        .unwrap()
-                        .unwrap_or(us.workload.workload_ip);
+                        .waypoint_svc_ip_address()?
+                        .unwrap_or(self.choose_workload_ip(&us.workload)?);
                     SocketAddr::from((ip, hbone_port))
                 }
-                Protocol::TCP => SocketAddr::from((us.workload.workload_ip, us.port)),
+                Protocol::TCP => {
+                    SocketAddr::from((self.choose_workload_ip(&us.workload)?, us.port))
+                }
             });
         }
+        Ok(())
+    }
+
+    // TODO: add more sophisticated routing logic, perhaps based on ipv4/ipv6 support underneath us.
```

Comment:
+1, randomly picking between ipv4 and ipv6 is .. weird. But no problem for now, we don't actually support dual stack yet